merkle-tree/concurrent/tests/tests.rs

```diff
@@ -0,0 +1,477 @@
+
+    let root_index = merkle_tree.root_index().unwrap();
+    merkle_tree
+        .update(root_index, &leaf1, &new_leaf1, 0, proof)
+        .unwrap();
+
+    let h1 = H::hashv(&[&new_leaf1, &leaf2]).unwrap();
+    let h2 = H::hashv(&[&leaf3, &leaf4]).unwrap();
+    let h3 = H::hashv(&[&h1, &h2]).unwrap();
+    let h4 = H::hashv(&[&h3, &H::zero_bytes()[2]]).unwrap();
+    let expected_root = H::hashv(&[&h4, &H::zero_bytes()[3]]).unwrap();
+
+    assert_eq!(merkle_tree.root(), expected_root);
+
+    // Replace `leaf2`.
+    let new_leaf2 = H::hash(&[8u8; 32]).unwrap();
+
+    // Merkle proof for the replaced leaf L2 is:
+    // [L1, H2]
+    let proof = &[new_leaf1, h2];
+
+    let root_index = merkle_tree.root_index().unwrap();
+    merkle_tree
+        .update(root_index, &leaf2, &new_leaf2, 1, proof)
+        .unwrap();
+
+    let h1 = H::hashv(&[&new_leaf1, &new_leaf2]).unwrap();
+    let h2 = H::hashv(&[&leaf3, &leaf4]).unwrap();
+    let h3 = H::hashv(&[&h1, &h2]).unwrap();
+    let h4 = H::hashv(&[&h3, &H::zero_bytes()[2]]).unwrap();
+    let expected_root = H::hashv(&[&h4, &H::zero_bytes()[3]]).unwrap();
+
+    assert_eq!(merkle_tree.root(), expected_root);
+
+    // Replace `leaf3`.
+    let new_leaf3 = [7u8; 32];
+
+    // Merkle proof for the replaced leaf L3 is:
+    // [L4, H1]
+    let proof = &[leaf4, h1];
+
+    let root_index = merkle_tree.root_index().unwrap();
+    merkle_tree
+        .update(root_index, &leaf3, &new_leaf3, 2, proof)
+        .unwrap();
+
+    let h1 = H::hashv(&[&new_leaf1, &new_leaf2]).unwrap();
+    let h2 = H::hashv(&[&new_leaf3, &leaf4]).unwrap();
+    let h3 = H::hashv(&[&h1, &h2]).unwrap();
+    let h4 = H::hashv(&[&h3, &H::zero_bytes()[2]]).unwrap();
+    let expected_root = H::hashv(&[&h4, &H::zero_bytes()[3]]).unwrap();
+
+    assert_eq!(merkle_tree.root(), expected_root);
+
+    // Replace `leaf4`.
+    let new_leaf4 = H::hash(&[6u8; 32]).unwrap();
+
+    // Merkle proof for the replaced leaf L4 is:
+    // [L3, H1]
+    let proof = &[new_leaf3, h1];
+
+    let root_index = merkle_tree.root_index().unwrap();
+    merkle_tree
+        .update(root_index, &leaf4, &new_leaf4, 3, proof)
+        .unwrap();
+
+    let h1 = H::hashv(&[&new_leaf1, &new_leaf2]).unwrap();
+    let h2 = H::hashv(&[&new_leaf3, &new_leaf4]).unwrap();
+    let h3 = H::hashv(&[&h1, &h2]).unwrap();
+    let h4 = H::hashv(&[&h3, &H::zero_bytes()[2]]).unwrap();
+    let expected_root = H::hashv(&[&h4, &H::zero_bytes()[3]]).unwrap();
+
+    assert_eq!(merkle_tree.root(), expected_root);
+}
+
+fn overfill_tree<H>()
+where
+    H: Hasher,
+{
+    const HEIGHT: usize = 2;
+    const ROOTS: usize = 32;
+
+    let mut merkle_tree = ConcurrentMerkleTree::<H, HEIGHT, ROOTS>::default();
+    merkle_tree.init();
+
+    for _ in 0..4 {
+        merkle_tree.append(&[4; 32]).unwrap();
+    }
+    assert!(matches!(
+        merkle_tree.append(&[4; 32]),
+        Err(HasherError::TreeFull)
+    ));
+}
+
+fn overfill_changelog<H>()
```

Comment:
Do you assert the values of the changelog somewhere?

Reply:
I don't, good idea to do so.